Telegram bot

Lfdd opened the pull request:
Forgot to merge TelegramBOT  branch

Lfdd committed:
Merge branch 'Documentation' into main

Lfdd committed:
Crated BOT based on aiogram module

Lfdd committed:
Update bot.py

Lfdd committed:
Modified to a working BOT for 1 author

Lfdd committed:
Merge remote-tracking branch 'origin/TelegramBot' into TelegramBot

# Conflicts:
#	elibrary_parser/bot.py

Lfdd committed:
Completed basic functions of bot. Added func. in 'utils'

Lfdd committed:
To Excel saving changes. Minor bot stability improvements

VladimirShitov commented on download_publications.py:
```diff
@@ -0,0 +1,28 @@
+from elibrary_parser.Parsers import AuthorParser
```
Может назовём этот файл `example.py` тогда? Чтобы было понятно, что он для примера

VladimirShitov commented on elibrary_parser/Parsers.py:
```diff
@@ -12,7 +12,7 @@
 
 from elibrary_parser import config
-from elibrary_parser.types import Publication
+from elibrary_parser.publicationwrk import Publication
```
А зачем здесь смена названия? Нормально же звучало. `publicationwrk` не очень понятно, что означает

Lfdd replied:
Был какой-то конфликт кажется. Нужно перепроверить

VladimirShitov commented on elibrary_parser/Parsers.py:
```diff
@@ -188,7 +188,9 @@ def save_publications(self):
         save_path.mkdir(exist_ok=True)
 
-        csv_path = save_path / "publications.csv"
+        authors_ids_csv = str(self.author_id) + "_publications.csv"
```
Можно 
```suggestion
        authors_ids_csv = f"{self.author_id)}_publications.csv"
```

Это чуть быстрее выполняется и читается, а суть такая же :)

VladimirShitov commented on elibrary_parser/Parsers.py:
```diff
@@ -188,7 +188,9 @@ def save_publications(self):
+
+        csv_path = save_path / authors_ids_csv
 
         with open(csv_path, 'a', encoding="utf8", newline='') as csvfile:
```
Определись с кавычками. У тебя в одних местах двойные, в других — одинарные. Правильного способа тут нет, просто реши, какой тебе больше нравится и используй везде его :)

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+
+def is_date(message: types.Message):
+    try:
+        if len(message.text) == 4 and int(message.text) and int(message.text) > 1900:
```
Не нужно писать
```python 
if something:
    return True
else:
    return False
```

Можно просто написать `return something` :)

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+    return True
+
+
+def is_date(message: types.Message):
```
Переменуй в `is_date_correct` лучше, наверное. Ещё в питоне есть стандартная библиотека для работы с датами (`datetime`, вроде), можно её использовать. Она лучше справится с валидацией дат

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+    await Form.next()
+    print(message.text.split())
+    async with state.proxy() as data:
+        data['authors_ids'] = (message.text.split())
```
Скобки вокруг выражения справа лишние :) А ещё, отбивай пустую строку до и после всяких вложенных конструкций (типа `with`, `for` и других, где будет отступ. Так читать проще

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+    action = call.data.split("_")[1]
+    if action == "extra":
+        await call.message.edit_text(f'Хорошо, введите новые id авторов через пробел')
+        await Form.authors_ids.set()
```
Здесь тоже отбивай пустыми строками блоки кода. Очень тяжело читать сплошной текст

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+async def callbacks(call: types.CallbackQuery):
+    action = call.data.split("_")[1]
+    if action == "extra":
+        await call.message.edit_text(f'Хорошо, введите новые id авторов через пробел')
```
Здесь `f` не обязателен. Он нужен только если ты вызываешь какой-то код внутри строки

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+            parser.parse_publications()  # Извлечение информации из HTML-файлов
+            parser.save_publications()  # Сохранение информации в CSV-файл
+
+            path = r"C:/Users/SZ/PycharmProjects/Parser/data/processed/" + \
```
1. Не нужно `r` до строки
2. Это будет работать только у тебя, не нужно хардкодить пути

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+            path = r"C:/Users/SZ/PycharmProjects/Parser/data/processed/" + \
+                (data['authors_ids']) + '/' + (data['authors_ids']) + r'_publications.csv'
+
+            await bot.send_document(message.chat.id, open(path, 'rb'))
```
Файл не закрывается после чтения. Нужно сделать так:
```python
with open(path, "rb") as f:
    await bot.send_document(message.chat.id, f)
```

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+            await bot.send_document(message.chat.id, open(path, 'rb'))
+
+        if len((data['authors_ids']).split()) > 1:
+            await message.reply("Ухх, я запустился! А должен был?")
```
Это весело, но давай такое в логи, а не пользователю отправлять :D

Lfdd replied:
Тестовая версия попала на ревью. Сорре

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+        if len((data['authors_ids']).split()) > 1:
+            await message.reply("Ухх, я запустился! А должен был?")
+
+            path_common = r"C:/Users/SZ/PycharmProjects/Parser/data/common_publications/" + \
```
Тоже хардкод пути, не будет работать на других компьютерах

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+            path_common = r"C:/Users/SZ/PycharmProjects/Parser/data/common_publications/" + \
+                (data['date_from']) + '-' + (data['date_to']) + r'_publications.csv'
+
+            data_path = "C://Users//SZ//PycharmProjects//Parser//data"
```
Ну ты понял

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+                (data['date_from']) + '-' + (data['date_to']) + r'_publications.csv'
+
+            data_path = "C://Users//SZ//PycharmProjects//Parser//data"
+            publications = [set(parser.publications)]
```
Тут не должно быть `list(set(parser.publications))`?

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+            save_common_publications(data_path=data_path, date_from=int(data['date_from']),
+                                     date_to=int(data['date_to']), publications=publications)
+
+            await bot.send_document(message.chat.id, open(path_common, 'rb'))
```
Тоже не закрывается файл

VladimirShitov commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+            await bot.send_document(message.chat.id, open(path_common, 'rb'))
+            await message.answer("Общие")
+        else:
+            await message.answer(f'Спасибо, что воспользовались нашим ботом!')
```
Не нужна форматируемая строка, убери `f` :)

VladimirShitov commented on elibrary_parser/utils.py:
```diff
@@ -1,2 +1,22 @@
+import pandas as pd
+
+
+def save_common_publications(data_path, date_from, date_to, publications):
```
Здесь смешались 2 функции: для поиска общих публикаций и для их сохранения. Нужно их разделить

VladimirShitov replied:
Раньше хорошая же функция была для поиска общих пубилкаций :) Мы можем захотеть ей ещё где-нибудь воспользоваться

VladimirShitov reviewed (changes requested):
Некоторые вещи, вроде закрытия файлов и захардкоженных путей обяазтельно нужно поправить. И сделай, пожалуйста, инструкцию к запуску в README, чтобы можно было протестировать у себя

IvanZavaruev commented on elibrary_parser/bot.py:
```diff
@@ -0,0 +1,156 @@
+def consists_of_integers(message: types.Message):
+    for i in message.text.split():
+        try:
+            if int(i) and int(i) > 0:
```
Честно говоря не понимаю, зачем проверять, что i  больше 0, как будет работать id с номером 100 000 (гипотетический) ? он же вернет false ?  
и не совсем понимаю сложную схему возвратов - буду благодарен за пояснения)

Lfdd committed:
Update elibrary_parser/Parsers.py

Co-authored-by: Vladimir Shitov <35199218+VladimirShitov@users.noreply.github.com>

Lfdd committed:
Update bot.py

Lfdd committed:
Delete utils.py

Это моя неудачная попытка сделать CSV в XLSX для нашего парсера. Можно удалить, т.к. в других ветках есть более подходящая реализация

Lfdd committed:
Update bot.py

Lfdd committed:
Rename publicationwrk.py to types.py

Lfdd committed:
Create utils.py

Lfdd committed:
Delete wrkwdata directory

Lfdd committed:
Create test_author_parser.py

Lfdd committed:
Update test_author_parser.py

Lfdd committed:
Create 1_download_publications.py

Lfdd committed:
Rename 1_download_publications.py to example.py

Lfdd reviewed (commented):
Комментарии учтены, бот работает

Lfdd committed:
Update Parsers.py
